### PR TITLE
BUG: fix handling of project name in package metadata

### PR DESCRIPTION
**Pull request by dnicolodi**
Just prototyping the solution for #481. More changes are needed to actually fix the bug, but these would be easier after #480 is merged, or at least merging the other PR first would avoid merge conflicts. I believe that none of the current users of meson-python exposes the bug, thus this is not really high priority yet.

### CODE REVIEW

**Comment by dnicolodi** [comment]
@rgommers This is ready for review

**Review by rgommers — APPROVED** [review]
Thanks @dnicolodi, this looks quite good to me. The test cleanups are nice too. I had one questions about the name validation method, which could perhaps save a bit of code. If you considered that already and that didn't seem like an improvement, then this can be merged as is.

File: mesonpy/__init__.py [hunk 1]
```diff
@@ -245,12 +246,26 @@ class MesonBuilderError(Error):
...
-    # The class method from the pyproject_metadata base class is not
-    # typed in a subclassing friendly way, thus annotations to ignore
-    # typing are needed.
+    def __init__(self, name: str, *args: Any, **kwargs: Any):
+        super().__init__(name, *args, **kwargs)
+        # Local fix for https://github.com/FFY00/python-pyproject-metadata/issues/60
+        self.name = self._validate_name(name)
+
+    @staticmethod
+    def _validate_name(name: str) -> str:
```
[the remark this answers is not in the document]
Comment (dnicolodi) [reply mesonpy/__init__.py:255]: The `validate` parameter to `packaging.utils.canonicalize_name()` has been added in `packaging` version 23.2 (looking this up now I discover that the function name reported in the `packaging` chagelog is wrong as it mentions an `utils.validate_name()` function that does not exist). I didn't want to bump the required `packaging` version for a regular expression match.

[the remark this answers is not in the document]
Comment (rgommers) [reply mesonpy/__init__.py:255]: Not bumping the minimum required version sounds like a good reason, thanks. In it goes then.

**Comment by rgommers** [comment]
We've accumulated several bug fixes and small new features for 0.15.0, and the PR backlog is pretty empty. WDYT about tagging a release now?

**Pull request merged**